### PR TITLE
improve: add gas tests for voting v1 and voting v2

### PR DESCRIPTION
**Pull request by mrice32**
**Motivation**

We would like to compare the gas costs of Voting and VotingV2.

**Summary**

This PR adds two test suites, one for Voting and one for VotingV2 that print out detailed gas costs for different methods.

Notes: 

- Because these print out lots of data, they are skipped unless the user provides the GAS_TEST=true environment variable.
- Because gas costs on the first run are often not representative (init costs), these scripts each do two runs. We should use the second for comparisons.
- This script commits and reveals individually since the batch methods are being removed. Once the multicall methods are in, we should swap these scripts to using batch in v1 and multicall in v2.

From the estimates below, it looks like the new voting system is marginally more expensive than the old one (1.03M gas for a 4 request lifecycle in VotingV2 vs 935k in Voting).

**Details**

Sample printout:

```
> GAS_TEST=true yarn hardhat test test/oracle/VotingGas.js test/oracle/VotingV2Gas.js
yarn run v1.22.5
$ /Users/matt/git/protocol/node_modules/.bin/hardhat test test/oracle/VotingGas.js test/oracle/VotingV2Gas.js

  Voting Gas Test


Initialize Run. Gas costs will likely be higher than typical

Gas costs for 4 voters, 4 requests

  Request Price
    
    First: 174389
    Last: 177189
    Max: 177189
    Min: 174389
    Average: 176489
    Average Per Voter: 176489
    Average Per Request: 176489
    Total For All Voters: 705956
    
  Commit
    
    First: 63660
    Last: 63660
    Max: 63660
    Min: 63648
    Average: 63659.25
    Average Per Voter: 254637
    Average Per Request: 254637
    Total For All Voters: 1018548
    
  Reveal
    
    First: 148223
    Last: 91463
    Max: 148223
    Min: 91463
    Average: 105653
    Average Per Voter: 422612
    Average Per Request: 422612
    Total For All Voters: 1690448
    
  RetrieveRewards
    
    First: 418487
    Last: 236900
    Max: 418487
    Min: 236888
    Average: 282293.75
    Average Per Voter: 282293.75
    Average Per Request: 282293.75
    Total For All Voters: 1129175
    
  Get Price
    
    First: 87114
    Last: 87114
    Max: 87114
    Min: 87114
    Average: 87114
    Average Per Voter: 87114
    Average Per Request: 87114
    Total For All Voters: 348456
    
  Total Lifecycle (only voting methods)
    
    First: 63660
    Last: 236900
    Max: 418487
    Min: 63648
    Average: 106615.86111111111
    Average Per Voter: 959542.75
    Average Per Request: 959542.75
    Total For All Voters: 3838171
    


Steady state run. Gas costs should be normal.

Gas costs for 4 voters, 4 requests

  Request Price
    
    First: 174389
    Last: 177189
    Max: 177189
    Min: 174389
    Average: 176489
    Average Per Voter: 176489
    Average Per Request: 176489
    Total For All Voters: 705956
    
  Commit
    
    First: 63660
    Last: 63660
    Max: 63660
    Min: 63648
    Average: 63658.5
    Average Per Voter: 254634
    Average Per Request: 254634
    Total For All Voters: 1018536
    
  Reveal
    
    First: 151391
    Last: 94631
    Max: 151391
    Min: 94631
    Average: 108821
    Average Per Voter: 435284
    Average Per Request: 435284
    Total For All Voters: 1741136
    
  RetrieveRewards
    
    First: 361937
    Last: 206313
    Max: 361937
    Min: 206301
    Average: 245216
    Average Per Voter: 245216
    Average Per Request: 245216
    Total For All Voters: 980864
    
  Get Price
    
    First: 87114
    Last: 87114
    Max: 87114
    Min: 87114
    Average: 87114
    Average Per Voter: 87114
    Average Per Request: 87114
    Total For All Voters: 348456
    
  Total Lifecycle (only voting methods)
    
    First: 63660
    Last: 206313
    Max: 361937
    Min: 63648
    Average: 103903.77777777778
    Average Per Voter: 935134
    Average Per Request: 935134
    Total For All Voters: 3740536
    
    ✓ Four voter gas test

  VotingV2 Gas Test


Initialize Run. Gas costs will likely be higher than typical

Gas costs for 4 voters, 4 requests

  Request Price
    
    First: 244971
    Last: 230671
    Max: 244971
    Min: 230671
    Average: 234246
    Average Per Voter: 234246
    Average Per Request: 234246
    Total for all voters: 936984
    
  Commit
    
    First: 213044
    Last: 108157
    Max: 213044
    Min: 108145
    Average: 121797.6875
    Average Per Voter: 487190.75
    Average Per Request: 487190.75
    Total for all voters: 1948763
    
  Reveal
    
    First: 135554
    Last: 78795
    Max: 135554
    Min: 78795
    Average: 92984.75
    Average Per Voter: 371939
    Average Per Request: 371939
    Total for all voters: 1487756
    
  Get Price
    
    First: 64514
    Last: 64514
    Max: 64514
    Min: 64514
    Average: 64514
    Average Per Voter: 64514
    Average Per Request: 64514
    Total for all voters: 258056
    
  Total lifecycle (only voting methods)
    
    First: 213044
    Last: 78795
    Max: 213044
    Min: 78795
    Average: 107391.21875
    Average Per Voter: 859129.75
    Average Per Request: 859129.75
    Total for all voters: 3436519
    


Steady state run. Gas costs should be normal.

Gas costs for 4 voters, 4 requests

  Request Price
    
    First: 230671
    Last: 230671
    Max: 230671
    Min: 230671
    Average: 230671
    Average Per Voter: 230671
    Average Per Request: 230671
    Total for all voters: 922684
    
  Commit
    
    First: 614685
    Last: 108157
    Max: 614685
    Min: 108145
    Average: 164925.5625
    Average Per Voter: 659702.25
    Average Per Request: 659702.25
    Total for all voters: 2638809
    
  Reveal
    
    First: 135566
    Last: 78807
    Max: 135566
    Min: 78807
    Average: 92996.75
    Average Per Voter: 371987
    Average Per Request: 371987
    Total for all voters: 1487948
    
  Get Price
    
    First: 64514
    Last: 64514
    Max: 64514
    Min: 64514
    Average: 64514
    Average Per Voter: 64514
    Average Per Request: 64514
    Total for all voters: 258056
    
  Total lifecycle (only voting methods)
    
    First: 614685
    Last: 78807
    Max: 614685
    Min: 78807
    Average: 128961.15625
    Average Per Voter: 1031689.25
    Average Per Request: 1031689.25
    Total for all voters: 4126757
    
    ✓ Four voter gas test
```

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [x]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

N/A

### CODE REVIEW

**Comment by amateima** [comment]
I'm just curious whats the biggest advantage of this way of measuring gas usage instead of using `hardhat-gas-reporter`

File: packages/core/test/oracle/VotingGas.js [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+Min: ${min}
+Average: ${average}
+Average Per Voter: ${costPerVoter}
+Average Per Request: ${costPerRequest}
+Total For All Voters: ${sum}
+`);
+    console.groupEnd();
+  }
+
+  async function runVoteAndPrintCosts(numRequests, numVoters, startTime) {
```
Comment (md0x) [inline packages/core/test/oracle/VotingGas.js:93]: This is very flexible, well thought out!

**Review by md0x — APPROVED** [review]
LGTM

**Comment by mrice32** [comment]
> I'm just curious whats the biggest advantage of this way of measuring gas usage instead of using `hardhat-gas-reporter`

@amateima the point of this sort of gas testing is to _control_ what you're testing and reduce the error bars. Typical gas reports show wild variations between the max and the min for functions.

Typical test runs have a lot of biases that are easy to miss:
- Most tests cases run against a fresh contract, so the gas costs will often be higher than steady state due to initializing variables from 0 -> nonzero.
- Similar to ^, gas costs are impacted by the current state/previous state of the contract. For instance, in VotingV2, if you only run a single voting cycle, you never run through a slashing tracker update (which would happen on the _second_ commit), so you would undercount the gas costs.
- You might want to test _specific scenarios_, like how the costs are affected by the number of requests or the number of voters, how much more the first voter pays than the second voter, etc.

You could control all of this in a particular test run, but still use the hardhat reporter _to print out_ the gas costs. However, then you lose the ability to aggregate different calls together (like the full voting lifecycle) or create your own metrics (like how I separated out the first call from the others).

This can all really be summarized as control. You have a lot more control over the assumptions you're using when you compare so you can ensure you're comparing "apples to apples".

Does that make sense?

**Review by chrismaree — APPROVED** [review]
this is awesome. thanks for adding it! keen to see after multicall is added

**Comment by amateima** [comment]
> > I'm just curious whats the biggest advantage of this way of measuring gas usage instead of using `hardhat-gas-reporter`
> 
> @amateima the point of this sort of gas testing is to _control_ what you're testing and reduce the error bars. Typical gas reports show wild variations between the max and the min for functions.
> 
> Typical test runs have a lot of biases that are easy to miss:
> 
> * Most tests cases run against a fresh contract, so the gas costs will often be higher than steady state due to initializing variables from 0 -> nonzero.
> * Similar to ^, gas costs are impacted by the current state/previous state of the contract. For instance, in VotingV2, if you only run a single voting cycle, you never run through a slashing tracker update (which would happen on the _second_ commit), so you would undercount the gas costs.
> * You might want to test _specific scenarios_, like how the costs are affected by the number of requests or the number of voters, how much more the first voter pays than the second voter, etc.
> 
> You could control all of this in a particular test run, but still use the hardhat reporter _to print out_ the gas costs. However, then you lose the ability to aggregate different calls together (like the full voting lifecycle) or create your own metrics (like how I separated out the first call from the others).
> 
> This can all really be summarized as control. You have a lot more control over the assumptions you're using when you compare so you can ensure you're comparing "apples to apples".
> 
> Does that make sense?

Yes. Thank you 🙌

**Pull request merged**